src/profilers/wall-profiler.ts

```diff
@@ -16,27 +16,63 @@ export interface WallProfilerStartArgs {
   collectCpuTime: boolean
 }
 
+export interface GenerateTimeLabelsArgs {
+  node: TimeProfileNode
+  context?: TimeProfileNodeContext
+}
+
+export interface TimeProfileNodeContext {
+  context: ProfilerContext
+  timestamp: bigint
+  cpuTime: number
+}
+
+export interface ProfilerContext {
+  labels?: LabelSet
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type AnyWrappedCallback = any
+
 export class WallProfiler implements Profiler<WallProfilerStartArgs> {
-  private labels: Record<string, number | string>
   private lastProfiledAt: Date
+  private lastContext: ProfilerContext
   private lastSamplingIntervalMicros!: number
 
   constructor() {
-    this.labels = {}
+    this.lastContext = {}
     this.lastProfiledAt = new Date()
   }
 
-  public getLabels(): Record<string, number | string> {
-    return this.labels
+  public getLabels(): LabelSet {
+    return this.lastContext.labels ?? {}
   }
 
   public profile(): ProfileExport {
     log('profile')
     return this.innerProfile(true)
   }
 
-  public setLabels(labels: Record<string, number | string>): void {
-    this.labels = labels
+  public wrapWithLabels(
+    lbls: LabelSet,
+    fn: () => void,
+    ...args: unknown[]
+  ): void {
+    const oldLabels = this.getLabels()
+    this.setLabels({
+      ...oldLabels,
+      ...lbls,
+    })
+    ;(fn as AnyWrappedCallback)(...args)
```

Comment:
Is this semicolon necessary?

```suggestion
    (fn as AnyWrappedCallback)(...args)
```

Reply:
I think so, don't ask me why though

![image](https://github.com/grafana/pyroscope-nodejs/assets/223048/38c9df30-3f82-43f6-8ea1-f4d2ca1821b0)

Reply:
Yes it is; it separates the statements.

If not, the JS engine tries to use the return value of `setLabels` as function that will receives the `fn` parameters.

Adding or not semicolons at the end of each line of JS code can be debated.
But for sure, not having them can create this confusion. Maybe we should configure Prettier/ESLint to format the code automatically (in a future PR).